### PR TITLE
Save test predictions to file

### PR DESCRIPTION
**Pull request by lokijuhy**
* Saves the predictions made during model training and testing to `<wrk_dir>/predictions.csv`. 
  * The csv contains document id, true class, predicted class, and log probabilities for each class prediction. 
  * In addition to assisting with deugging and identifying example predictions, this data is used for simulating accuracy scores with at different levels of coverage (only making predictions when the model's prediction confidence is above a certain threshold). 
* Switches `hyperparam-search-dir` command line argument for `experiment-to-rerun`, which uses the config from that experiment's `train_validation question_*/fold_*/config/` directories (instead of repeating the search for best hyperparam search dir).

### CODE REVIEW

File: neural/run_workflow.py [hunk 1]
```diff
@@ -324,6 +327,7 @@ def run_neural_discern(data_partition, dsettypes, bertmodel, config, options, wr
                     # finished processing docs in batch
                     b_logprob_scores = torch.cat(logprob_scores, dim=0)
                     b_target_class = torch.cat(target_class, dim=0)
+                    all_probability_scores.extend(logprob_scores)
```
Comment (orisenbazuru) [inline neural/run_workflow.py]: This one would turn into a problem (especially if you have more samples/data) as you are saving in the list the data and the whole computation graph. One option is to do like this `all_probability_scores.extend(b_logprob_scores.tolist())`. You will get a list of list (i.e. [[logprob0_i, logprob1_i],...]) without saving the tensor or the history of graph computation etc..
If you update it then `build_predictions_df(ids, true_class, pred_class, probability_tensor)` needs to be updated. You will simply iterate over the list and just take the first and second element consecutively.

File: neural/run_workflow.py [hunk 2]
```diff
@@ -372,9 +376,30 @@ def run_neural_discern(data_partition, dsettypes, bertmodel, config, options, wr
...
+    # save predictions
+    predictions_df = build_predictions_df(doc_ids, ref_class, pred_class, all_probability_scores)
+    predictions_path = os.path.join(wrk_dir, 'predictions.csv')
+    predictions_df.to_csv(predictions_path)
+
     return pred_class
 
 
+def build_predictions_df(ids, true_class, pred_class, probability_tensor):
+    class_0_score = [t[0].item() for t in probability_tensor.cpu()]
```
Comment (orisenbazuru) [inline neural/run_workflow.py]: If you update how you are saving `logprobscores` then `probability_tensor` will be a list of lists. For example, `class_0_score=[[t[0] for t in probability_tensor]`. Maybe renaming `probability_tebnsor` into `logproba_list` :)

File: neural/run_workflow.py [hunk 3]
```diff
@@ -216,8 +217,7 @@ def run_neural_discern(data_partition, dsettypes, bertmodel, config, options, wr
     docid_attnweights_map = {dsettype: {} for dsettype in data_loaders if dsettype in {'validation', 'test'}}
     # store sentences' attention weights
 
-    if('validation' in data_loaders):
-        m_state_dict_dir = create_directory(os.path.join(wrk_dir, 'model_statedict'))
+    m_state_dict_dir = create_directory(os.path.join(wrk_dir, 'model_statedict'))
```
Comment (orisenbazuru) [inline neural/run_workflow.py]: currently you create "model_statedict" directory by default. Typically, you need that when you are training and validating on a validation set. It is not that important here -- that's why there is an `if` statement to check the case of `training and validation`.

**Review by orisenbazuru — CHANGES_REQUESTED** [review]
review with comments.

**Review by orisenbazuru — APPROVED** [review]
